drivers/gpio/gpio_adp5585.c

```diff
@@ -0,0 +1,478 @@
+{
+	const struct adp5585_gpio_config *cfg = dev->config;
+	struct adp5585_gpio_data *data = dev->data;
+	const struct mfd_adp5585_config *parent_cfg =
+		(struct mfd_adp5585_config *)(cfg->mfd_dev->config);
+	struct mfd_adp5585_data *parent_data = (struct mfd_adp5585_data *)(cfg->mfd_dev->data);
+
+	int ret = 0;
+	uint8_t reg_value;
+
+	/* ADP5585 has non-contiguous gpio pin layouts, account for this */
+	if ((pin & cfg->common.port_pin_mask) == 0) {
+		LOG_ERR("pin %d is invalid for this device", pin);
+		return -ENOTSUP;
+	}
+
+	uint8_t bank = ADP5585_BANK(pin);
+	uint8_t bank_pin = ADP5585_BIT(pin);
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	/* Single Ended lines (Open drain and open source) not supported */
+	if ((flags & GPIO_SINGLE_ENDED) != 0) {
+		return -ENOTSUP;
+	}
+
+	/* Simultaneous PU & PD mode not supported */
+	if (((flags & GPIO_PULL_UP) != 0) && ((flags & GPIO_PULL_DOWN) != 0)) {
+		return -ENOTSUP;
+	}
+
+	/* Simultaneous input & output mode not supported */
+	if (((flags & GPIO_INPUT) != 0) && ((flags & GPIO_OUTPUT) != 0)) {
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&parent_data->lock, K_FOREVER);
+
+	if ((flags & GPIO_SINGLE_ENDED) != 0) {
+		reg_value = adp5585_pin_drive_od << bank_pin;
+	} else {
+		reg_value = adp5585_pin_drive_pp << bank_pin;
+	}
+	ret = i2c_reg_update_byte_dt(&parent_cfg->i2c_bus, ADP5585_GPO_OUT_MODE_A + bank,
+				BIT(bank_pin), reg_value);
+	if (ret != 0) {
+		goto out;
+	}
+
+	{
+		uint8_t regaddr = ADP5585_RPULL_CONFIG_A + (bank << 1);
+		uint8_t shift = bank_pin << 1;
+
+		if (bank_pin > 3U) {
+			regaddr += 1U;
+			shift = (bank_pin - 3U) << 1;
+		}
+		if ((flags & GPIO_PULL_UP) != 0) {
+			reg_value = adp5585_pull_up_300k << shift;
+		} else if ((flags & GPIO_PULL_DOWN) != 0) {
+			reg_value = adp5585_pull_dn_300k << shift;
+		} else {
+			reg_value = adp5585_pull_disable << shift;
+		}
+
+		ret = i2c_reg_update_byte_dt(&parent_cfg->i2c_bus, regaddr,
+				0b11U << shift, reg_value);
+		if (ret != 0) {
+			goto out;
+		}
+	}
+
+	/* Ensure either Output or Input is specified */
+	if ((flags & GPIO_OUTPUT) != 0) {
+
+		/* Set Low or High if specified */
+		if ((flags & GPIO_OUTPUT_INIT_LOW) != 0) {
+			data->output &= ~BIT(pin);
+		} else if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
+			data->output |= BIT(pin);
+		}
+		if (bank == 0) {
+			/* reg_value for ADP5585_GPO_OUT_MODE */
+			reg_value = (uint8_t)data->output;
+		} else {
+			/* reg_value for ADP5585_GPO_OUT_MODE */
+			reg_value = (uint8_t)(data->output >> 8);
+		}
+		ret = i2c_reg_write_byte_dt(&parent_cfg->i2c_bus,
+					ADP5585_GPO_OUT_MODE_A + bank,
+					reg_value);
+		if (ret != 0) {
+			goto out;
+		}
+		/* reg_value for ADP5585_GPIO_DIRECTION */
+		reg_value = adp5585_pin_output << bank_pin;
+	} else if ((flags & GPIO_INPUT) != 0) {
+		/* reg_value for ADP5585_GPIO_DIRECTION */
+		reg_value = adp5585_pin_output << bank_pin;
+	} else {
```

Comment:
This condition has an explicit check earlier

Reply:
sure, remove it here.